src/Processors/Formats/Impl/AvroRowInputFormat.cpp

```diff
@@ -592,19 +592,30 @@ void AvroDeserializer::deserializeRow(MutableColumns & columns, avro::Decoder &
 
 
 AvroRowInputFormat::AvroRowInputFormat(const Block & header_, ReadBuffer & in_, Params params_, const FormatSettings & format_settings_)
-    : IRowInputFormat(header_, in_, params_)
-    , file_reader(std::make_unique<InputStreamReadBufferAdapter>(in_))
-    , deserializer(output.getHeader(), file_reader.dataSchema(), format_settings_)
+    : IRowInputFormat(header_, in_, params_),
+      allow_missing_fields(format_settings_.avro.allow_missing_fields)
+{
+}
+
+void AvroRowInputFormat::resetParser()
 {
-    file_reader.init();
+    IRowInputFormat::resetParser();
+    file_reader_ptr.reset();
 }
 
 bool AvroRowInputFormat::readRow(MutableColumns & columns, RowReadExtension &ext)
 {
-    if (file_reader.hasMore())
+    if (!file_reader_ptr)
+    {
+        file_reader_ptr = std::make_unique<avro::DataFileReaderBase>(std::make_unique<InputStreamReadBufferAdapter>(in));
```

Comment:
Maybe better to do it in read prefix? To avoid extra if for every row?

Reply:
The goal is to read header and stuff when AvroRowInputFormat is constructed or when resetParser is called. Although we cannot do it in constructor (and might be in resetParser) context because data is not ready, it must be done in context of readRow. 
That is why we need some state.
We can invent more elegant and less straightforward approach, but plain 'if' seems to be a winner here.

Reply:
There is one more stage between constructor and readrow. It's called readPrefix and is called when starting reading the message / file (not for every row), see for example 
https://github.com/ClickHouse/ClickHouse/blob/d615b8e516569ddf69ad92cd3b73f6591c0b7248/src/Processors/Formats/Impl/CSVRowInputFormat.cpp#L149-L153

Reply:
Done.